transaction/core/src/range.rs

```diff
@@ -29,17 +29,21 @@ impl Range {
 
     /// The number of indices in this Range.
     pub fn len(&self) -> u64 {
-        self.to - self.from + 1
+        if self.from <= self.to {
+            self.to - self.from + 1
+        } else {
+            self.from - self.to + 1
+        }
     }
 }
 
-/// Ranges are ordered by `len`. If two ranges have equal `len`, they are ordered by `from`.
+/// Ranges are ordered by `len`, then lexicographically by (from,to).
```

Comment:
I feel like we may be mixing cardinality and ordinality here? I would have expected ordering by "from" and then by length (kind of like alphabetizing words by the first letter, and then the length)

Reply:
This ordering of Ranges has a nice property when a node of a binary tree (i.e. Merkle tree) are identified with the range of indices below that node: the `len` of a range is equal to 2^h, where h is the height of the node in the tree. This means that traversing a list of ranges sorted in ascending order corresponds to a bottom-up traversal of the tree (which is handy for computing Merkle hashes). Ordering ranges of equal `len` lexicographically makes it a bottom-up and left-to-right traversal of the tree. This ordering also handles the case where "from" may be greater than "to". I don't think sorting ("from", "len") is actually a total ordering here, because (5, 7) and (5, 3) both have "from" 5 and "len" 3 according to the definition of `len` here.

That^ is definitely not obvious, so I'll add it as a comment!